### ISSUE TITLE
None and false prediction score when evaluating my custom model using GoldParse

### ISSUE THREAD

**Issue by Todaime** [explosion/spaCy#5322]
Hello there,

After training my custom ner spacy model, i wanted to calculate the score of prediction but i get the same false result for each prediction : 

```
scorer = Scorer()
doc_gold_text = model.make_doc(labeled_data[1][0])
gold = GoldParse(doc_gold_text, labeled_data[1][1]['entities'])
pred_value = model(labeled_data[1][0])
scorer.score(pred_value, gold)
print(model.evaluate([(pred_value,gold)]).scores)
```
give me : 

`{'uas': 0.0, 'las': 0.0, 'las_per_type': {'': {'p': 0.0, 'r': 0.0, 'f': 0.0}}, 'ents_p': 0.0, 'ents_r': 0.0, 'ents_f': 0.0, 'ents_per_type': {}, 'tags_acc': 0.0, 'token_acc': 100.0, 'textcat_score': 0.0, 'textcats_per_cat': {}}`

Looking to the GoldParse Heads, i got a list of None : 
`print(gold.heads)`
`[None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, ...
`
My test input is  :

```
(text1,{'entities': [(0,31, 'Titre'),
                (98,106, 'Label'),(122,155, 'Valeur'),
                (193,210, 'Label'),(217,224, 'Valeur'),
                (291,295, 'Label'),(315,354, 'Valeur'),
                (422,428, 'Label'),(446,466, 'Valeur'),
                (504,520, 'Label'),(528,542, 'Valeur'),
                (608,623, 'Label'),(632,637, 'Valeur'),
                (675,687, 'Label'),(699,704, 'Valeur'),
                (768,785, 'Label'),(792,807, 'Valeur'),
                (845,860, 'Label'),(869,884, 'Valeur'),
                (954,975, 'Label'),(978,993, 'Valeur'),
                (1059,1074, 'Label'),(1083,1087, 'Valeur'),
                (1197,1203, 'Label'),(1221,1224, 'Valeur'),
                (3301,3323, 'Label'),(3325,3361, 'Valeur'),
                (3428,3445, 'Label'),(3452,3502, 'Valeur'),
                (4748,4760, 'Label'),(4772,4819, 'Valeur'),
                (8361,8370, 'Label'),(8384,8442, 'Valeur'),
                (8514,8521, 'Label'),(8538,8549, 'Valeur'),
                (8618,8627, 'Label'),(8641,8692, 'Valeur')]})
```
- Does anyone have an idea about that? 

- And I was wondering, is there a parameter during training so that the ner model change his weight according to a validation set?

## Your Environment

* Operating System: Darwin-19.3.0-x86_64-i386-64bit
* Python Version Used: 3.7.6
* spaCy Version Used: 2.2.4

**Comment by svlandeg** [comment]
What was the command you used to train, and the output of the training iterations?

**Comment by adrianeboyd** [comment]
In your eval code above, I think you need to use the `entities` argument for `GoldParse`:

```
gold = GoldParse(doc, entities=annotation['entities'])
```

`gold.heads` is part of the dependency parse and it doesn't look like you have dependency annotation in your data. For NER annotations, you want to check `gold.ner` to make sure the data was read in correctly. You might see `-` instead of BILUO tags if the character offsets don't line up with token boundaries.

**Comment by Todaime** [comment]
@svlandeg Thank you for your quick answer ! ;) 
 - To train my model I used : 

```
nlp = spacy.blank('xx')  # Creation d'un model blanc.

if 'ner' not in nlp.pipe_names :
        ner = nlp.create_pipe('ner')
        nlp.add_pipe(ner, last=True)
else :
        ner = nlp.get_pipe('ner')
 for _, annotations in data_set:
        for ent in annotations.get('entities'):
            ner.add_label(ent[2])

other_pipe = [pipe for pipe in nlp.pipe_names if pipe != 'ner']

with nlp.disable_pipes(*other_pipe) :
            optimizer = nlp.begin_training()

for ite in range(nb_iteration) :
        compteur_iteration += 1
        random.shuffle(train_set)
        losses = {}
        for text, annotation in data_set :
            nlp.update(
            [text],
            [annotation],
            drop = 0.3,
            sgd = optimizer,
            losses = losses
            )
```
So the output is the nlp trained model :) 

@adrianeboyd, I added the entities param name (it doesn't change the metric values) and gold.ner  give me that :

`['B-Titre', 'I-Titre', 'I-Titre', 'I-Titre', 'L-Titre', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'U-Label', 'O', 'B-Valeur', 'I-Valeur', 'I-Valeur', 'I-Valeur', 'I-Valeur', 'I-Valeur', 'I-Valeur', 'I-Valeur', 'L-Valeur', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'B-Label', 'L-Label', 'O', 'U-Valeur', 'O', 'O' 'O', 'O', 'O', 'O',....`

 Again , thank you for your time :)

Devly yours 💻

**Comment by adrianeboyd** [comment]
The `GoldParse` NER data looks good in the example above, but it's hard to know exactly what's going on in the training step without a few more details.

Can you make a complete example that trains and test on the same small set of sentences (with anonymized texts if you can't share the real data)? It should overfit and the loss should decrease quickly. Here's the closest example training script: https://github.com/explosion/spaCy/blob/f7471abd82c1cbf6dc42f299eb0237a174f86da5/examples/training/train_new_entity_type.py

It would be easy to run `nlp.evaluate()` after training in this example on the training data instead of showing the output for one sentence.

**Comment by Todaime** [comment]
Sorry, for the delay... 

I trained with the script you shared using only one of my sample as a training set ....

After 100 iterations , the ner Losse reach the value of 0.0011295166045127189. But the nlp score is still :
`{'uas': 0.0, 'las': 0.0, 'las_per_type': {'': {'p': 0.0, 'r': 0.0, 'f': 0.0}}, 'ents_p': 0.0, 'ents_r': 0.0, 'ents_f': 0.0, 'ents_per_type': {}, 'tags_acc': 0.0, 'token_acc': 100.0, 'textcat_score': 0.0, 'textcats_per_cat': {}}
`

Here is the code used : 
```
TRAIN_DATA = [
    ("text2",{'entities':[(0,14, 'Titre'),
                (15,21, 'Label'),(41,44, 'Valeur'),
                (45,57, 'Label'),(71,88, 'Valeur'),
                (89,99, 'Label'),(115,129, 'Valeur'),
                (130,143, 'Label'),(156,183, 'Valeur'),
                (184,195, 'Label'),(210,217, 'Valeur'),
                (218,252, 'Label'),(253,267, 'Valeur'),
                (1709,1726, 'Label'),(1735,1738, 'Valeur'),
                (1739,1752, 'Label'),(1765,1837, 'Valeur'),
                (1838,1850, 'Label'),(1864,1879, 'Valeur'),
                (1893,1905, 'Label'),(1919,1938, 'Valeur'),
                (1939,1947, 'Label'),(1965,1997, 'Valeur'),
                (1998,2008, 'Label'),(2024,2042, 'Valeur'),
                (2043,2051, 'Label'),(2069,2090, 'Valeur'),
                (2394,2402, 'Label'),(2420,2434, 'Valeur'),
                (2619,2635, 'Label'),(2645,2695, 'Valeur'),
                (2832,2845, 'Label'),(2858,2930, 'Valeur'),
                (2931,2948, 'Label'),(2957,2960, 'Valeur'),
                (2961,2973, 'Label'),(2987,3044, 'Valeur')]})]


@plac.annotations(
    model=("Model name. Defaults to blank 'xx' model.", "option", "m", str),
    new_model_name=("New model name for model meta.", "option", "nm", str),
    output_dir=("Optional output directory", "option", "o", Path),
    n_iter=("Number of training iterations", "option", "n", int),
)
def train(model=None, new_model_name="overfit", output_dir=None, n_iter=200):
    """Set up the pipeline and entity recognizer, and train the new entity."""
    random.seed(0)
    if model is not None:
        nlp = spacy.load(model)  # load existing spaCy model
        print("Loaded model '%s'" % model)
    else:
        nlp = spacy.blank("xx")  # create blank Language class
        print("Created blank 'xx' model")
    # Add entity recognizer to model if it's not in the pipeline
    # nlp.create_pipe works for built-ins that are registered with spaCy
    if "ner" not in nlp.pipe_names:
        ner = nlp.create_pipe("ner")
        nlp.add_pipe(ner)
    # otherwise, get it, so we can add labels to it
    else:
        ner = nlp.get_pipe("ner")

    ner.add_label("Label")  # add new entity label to entity recognizer
    ner.add_label("Valeur")  # add new entity label to entity recognizer
    ner.add_label("Titre")
    # Adding extraneous labels shouldn't mess anything up
    ner.add_label("VEGETABLE")
    if model is None:
        optimizer = nlp.begin_training()
    else:
        optimizer = nlp.resume_training()
    move_names = list(ner.move_names)
    # get names of other pipes to disable them during training
    pipe_exceptions = ["ner", "trf_wordpiecer", "trf_tok2vec"]
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe not in pipe_exceptions]
    with nlp.disable_pipes(*other_pipes):  # only train NER
        sizes = compounding(1.0, 4.0, 1.001)
        # batch up the examples using spaCy's minibatch
        for itn in range(n_iter):
            random.shuffle(TRAIN_DATA)
            batches = minibatch(TRAIN_DATA, size=sizes)
            losses = {}
            for batch in batches:
                texts, annotations = zip(*batch)
                nlp.update(texts, annotations, sgd=optimizer, drop=0.35, losses=losses)
            print("Losses", losses)
    print(nlp.evaluate(TRAIN_DATA).scores)
    # test the trained model
    test_text = "Do you like horses?"
    doc = nlp(test_text)
    print("Entities in '%s'" % test_text)
    for ent in doc.ents:
        print(ent.label_, ent.text)
    
    
    # save model to output directory
    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()
        nlp.meta["name"] = new_model_name  # rename model
        nlp.to_disk(output_dir)
        print("Saved model to", output_dir)

        # test the saved model
        print("Loading from", output_dir)
        nlp2 = spacy.load(output_dir)
        # Check the classes have loaded back consistently
        assert nlp2.get_pipe("ner").move_names == move_names
        doc2 = nlp2(test_text)
        for ent in doc2.ents:
            print(ent.label_, ent.text)
    return nlp
nlp = train()
```


Here is the text2 with modified value for confidentiality : 
```
New Issue Note
Issuer                ：   TdS
Issue Amount          ：   JPY 0,000,000,042
Issue Date            ：   April 12, 2712
Maturity Date         ：   April 30, 2732 (unadjusted)
Issue Price           ：   100.00%
Coupon Calculation Commencing Date：April 12, 2712
Coupon Payment Date:      every April 30 and September 30 starting from September 30, 2712. Payable semi-
                          annually in arrear.
Coupon Rate            ：  1yr                   13.23%
                          Thereafter            (FX1 – 18.20 ) x 6 x 9%
                          (Maximum Rate=19.32%, Minimum Rate = 3.21%, per annum, 23/360 unadjusted)
                          Coupon shall be rounded to the nearest yen with 1.23 yen being rounded up
FX1                   :   BID side of JPYZAR spot FX rate which appears on REUTERS ”JPNX” as of 12:22
                          Paris time on Coupon Determination Date.
Coupon Determination Date: 03 Business Days prior to every Coupon Payment Date.
Redemption            ：   (A) If the bond is early redeemed       JPY 112%
                          (B) If the bond remains until the Maturity Date (April 30, 2763)
                                                                      113.00% x FX2 / 13.00
            FX2     :      BID side of JPYZAR spot FX rate which appears on REUTERS ”JPNX” as of 12:31
            Paris time 38 Business days prior to the Maturity Date.
Early Redemption      ：   The issuer has right to call the note on every Coupon Payment Date from and
                          including, 30 April, 2737 to and including 24 September, 2765 (“Early Redemption
                          Date”) at par with 38 Business Days prior notice.
Calculation Agent     ：   xxx
Business Days         ：   Paris, London, Johannesburg, Barcelona and New York (Modified Following)
Denomination          ：   JPY 000,000,030
Hedging Swap
Counterparty            : Issuer (“CP”) / XXX
Notional                : JPY0,000,000,030 / USD 0,000,020
Start Date              : Issue Date of EMTN
End Date                : Maturity Date of EMTN
CP pays                 : 1y               USD 1m LIBOR – 48bps (long first without interpolation)
                          Thereafter       USD 1m LIBOR – 23bps
                          (quarterly, act/360, adjusted)
                           Roll Date : 30th of April, June, September and December
XXX pays                : Coupon on EMTN
Early Termination       : XXX has right to call the swap on every Early Redemption Date of above mentioned
                          EMTN with 11 Business Days prior notice to Issuer.
Initial Exchange        : Issuer pays JPY Notional and XXX pays USD Notional
Final Exchange          : Issuer pays USD Notional and XXX pays the Redemption Amount of above mentioned
                          EMTN
Business Days           : Paris, London, Johannesburg, Barcelona and New York (Modified Following)
Calculation Agent       : XXX
Mutual Break            : Mutual breaks on 30 April, 2726 and every year thereafter
```

For the record , the purpose of my work is to label an entire pdf document so each test sample contains like 100 up to 300 labels. I don't know if it impact the score.

**Comment by svlandeg** [comment]
Sorry for the very late follow-up ... Unfortunately I can't quite replicate your error.

I do think that this should be resolved in the latest [pre-release of spaCy v3](https://github.com/explosion/spaCy/releases/tag/v3.0.0rc1), where training should go through the [training CLI](https://nightly.spacy.io/usage/training), preventing issues with custom training scripts etc. We've also entirely revamped the internal constructs - `GoldParse` has now been replaced with `Example`, which makes things much more intuitive IMO. See for more docs here: https://nightly.spacy.io/api/data-formats#dict-input

I'll close this in the meantime, but if you do keep running into issues, feel free to open a new issue!
